internal/protoyaml/protoyaml.go

```diff
@@ -0,0 +1,907 @@
+	"github.com/goccy/go-yaml"
+	"github.com/goccy/go-yaml/ast"
+	"github.com/goccy/go-yaml/lexer"
+	"github.com/goccy/go-yaml/parser"
+	"github.com/goccy/go-yaml/printer"
+	"github.com/goccy/go-yaml/token"
+	"google.golang.org/protobuf/encoding/protowire"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	sourcev1 "github.com/cerbos/cerbos/api/genpb/cerbos/source/v1"
+)
+
+const (
+	base10    = 10
+	bitSize32 = 32
+	bitSize64 = 64
+)
+
+type unmarshalOpts struct {
+	validator           *protovalidate.Validator
+	fixInvalidStrings   bool
+	ignoreUnknownFields bool
+}
+
+type UnmarshalOpt func(*unmarshalOpts)
+
+// WithFixInvalidStrings makes the unmarshaler handle values that are partially quoted such as: "foo" bar.
+func WithFixInvalidStrings() UnmarshalOpt {
+	return func(uo *unmarshalOpts) {
+		uo.fixInvalidStrings = true
+	}
+}
+
+// WithIgnoreUnknownFields ignores unknown fields not defined in the protobuf schema.
+func WithIgnoreUnknownFields() UnmarshalOpt {
+	return func(uo *unmarshalOpts) {
+		uo.ignoreUnknownFields = true
+	}
+}
+
+// WithValidate validates the unmarshaled message using protovalidate.
+func WithValidator(validator *protovalidate.Validator) UnmarshalOpt {
+	return func(uo *unmarshalOpts) {
+		uo.validator = validator
+	}
+}
+
+type Unmarshaler[T proto.Message] struct {
+	factory func() T
+	anchors map[string]ast.Node
+	unmarshalOpts
+}
+
+func NewUnmarshaler[T proto.Message](factory func() T, opts ...UnmarshalOpt) *Unmarshaler[T] {
+	uo := unmarshalOpts{}
+	for _, o := range opts {
+		o(&uo)
+	}
+
+	return &Unmarshaler[T]{factory: factory, unmarshalOpts: uo}
+}
+
+func (u *Unmarshaler[T]) UnmarshalReader(r io.Reader) ([]T, []*sourcev1.SourceContext, error) {
+	contents, err := io.ReadAll(r)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to read contents: %w", err)
+	}
+
+	return u.unmarshal(contents)
+}
+
+func (u *Unmarshaler[T]) unmarshal(contents []byte) (_ []T, _ []*sourcev1.SourceContext, outErr error) {
+	t := lexer.Tokenize(unsafe.String(unsafe.SliceData(contents), len(contents)))
+	if u.fixInvalidStrings {
+		t = fixStringsStartingWithQuotes(t)
+	}
+
+	f, err := parser.Parse(t, 0)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(f.Docs) == 0 {
+		return nil, nil, nil
+	}
+
+	outMsg := make([]T, 0, len(f.Docs))
+	outSrc := make([]*sourcev1.SourceContext, 0, len(f.Docs))
+	for _, doc := range f.Docs {
+		msg := u.factory()
+		srcCtx := &sourcev1.SourceContext{FieldPositions: make(map[string]*sourcev1.Position)}
+		uctx := &unmarshalCtx{doc: doc, srcCtx: srcCtx}
+		if err := u.unmarshalDoc(uctx, doc, msg); err != nil {
+			outErr = errors.Join(outErr, err)
+		} else if err := u.validate(uctx, msg); err != nil {
+			outErr = errors.Join(outErr, err)
+		}
+
+		outMsg = append(outMsg, msg)
+		outSrc = append(outSrc, srcCtx)
+	}
+
+	return outMsg, outSrc, outErr
+}
+
+func fixStringsStartingWithQuotes(tokens token.Tokens) token.Tokens {
```

Comment:
I'm not so keen on this. By doing this, we're then defining our own policy language, it's no longer standard YAML. The SDKs won't be able to parse the same files that Cerbos can because they use vanilla YAML parsers.

Given it is a common error, I think it definitely makes sense to detect this case and offer a more helpful error message suggesting the use of alternative string literal syntaxes, but I don't think we should go as far as actually accepting invalid YAML.